### ISSUE TITLE
Software always hang when save file big than 100M

### ISSUE THREAD

**Issue by thinkport** [dokan-dev/dokany#719]
# Environment
 * Windows version:  Win7 x64/Win10 x64
 * Processor architecture: x64
 * Dokany version: 1.1.0.2000
 * Library type (Dokany/FUSE): Dokany/Mirror
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description below.
  - [x] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [ ] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

# Description
Software always hang when save file big than 100M, any software, any file.
 
I do more test in v1.0.0, v1.0.1, v1.0.2, v1.0.3, v1.0.4, v1.0.5, v1.1.0.1000, v2.0.0 have this bug too
# Logs

**Comment by thinkport** [comment]
Yes, dokan is able to open big file, but hang when I save big file, for example MS Office, WPS, AutoCAD, 3D Max. 

testcase
1. Open a .docx file using MS Office,  the file big than 100M
2. Change something
3. Press Ctrl + S to save file, Word.exe and explorer.exe hang

**Comment by Liryna** [comment]
@thinkport OK that's looks more like a lock issue. There à PR that should fix this issue for the next release coming next week.

**Comment by thinkport** [comment]
@Liryna Sorry, my English is not good. MS Office and explorer are fix, but WPS doesn't work.

change:
`//#define DokanFCBLockRO(fcb) do { KeEnterCriticalRegion(); ExAcquireResourceSharedLite(fcb->AdvancedFCBHeader.Resource, TRUE); } while(0)

//#define DokanFCBLockRW(fcb) ExEnterCriticalRegionAndAcquireResourceExclusive(fcb->AdvancedFCBHeader.Resource)

//#define DokanFCBUnlock(fcb) ExReleaseResourceAndLeaveCriticalRegion(fcb->AdvancedFCBHeader.Resource)

#define DokanFCBLockRO(fcb) do { DDbgPrint("ZZZ LockRO %s %p\n", __FUNCTION__, fcb); KeEnterCriticalRegion(); ExAcquireResourceSharedLite(fcb->AdvancedFCBHeader.Resource, TRUE); KeLeaveCriticalRegion(); } while(0)

#define DokanFCBLockRW(fcb) do { DDbgPrint("ZZZ LockRW %s %p\n", __FUNCTION__, fcb); KeEnterCriticalRegion(); ExAcquireResourceExclusiveLite(fcb->AdvancedFCBHeader.Resource, TRUE); KeLeaveCriticalRegion(); } while(0)

#define DokanFCBUnlock(fcb) do { DDbgPrint("ZZZ Unlock %s %p\n", __FUNCTION__, fcb); KeEnterCriticalRegion(); ExReleaseResourceLite(fcb->AdvancedFCBHeader.Resource); KeLeaveCriticalRegion(); } while(0)
`

DokanFS log:
[wps hang.LOG](https://github.com/dokan-dev/dokany/files/2249204/wps.hang.LOG)


WPS will hang every time when I save big file. WPS hang, i open notepad.exe, notepad hang too.  you can see,  DokanCreate never return
00439088	114.82792664	[DokanFS] ZZZ LockRO DokanGetFCB FFFFFA803289A050
If i remove fcb lock, wps never hang , but some content of the file are missing sometimes.

**Comment by Liryna** [comment]
Hi @thinkport ,

When you say fix, you mean with this PR https://github.com/dokan-dev/dokany/pull/632 that I was talking about ?

**Comment by thinkport** [comment]
@Liryna 
I am sorry, I am not sure, I can't reproduce MS office lock problem.

I can reproduce WPS lock problem always.  WPS can open MS office document.
download address [https://www.wps.com/download/?lang=en](url)

This deadlock problem is mapping file problem.

mirror.exe /r c: /l m:

Normal:  wps.exe(m:\bigfile.doc) -> dokan1.sys -> mirror.exe(c:\bigfile.doc) -> system(c:\bigfile.doc). work well always. 

Problem:  system(m:\bigfile.doc) -> dokan1.sys -> mirror.exe(c:\bigfile.doc). Last WriteFile function never return, you can see by Procmon.exe.

![system](https://user-images.githubusercontent.com/12952910/43888273-ce505abc-9bf3-11e8-83ac-797549b02f1b.png)

**Comment by thinkport** [comment]
The problem has not been solved. Autodesk 3ds Max and Siemens NX 11(Unigraphics NX, formerly known as "UG") have the same problem.
Like [WinFSP](https://github.com/billziss-gh/winfsp/issues/189)
